### ISSUE TITLE
Not working with multiprocessing?

### ISSUE THREAD

**Issue by rafi-cohen** [gaogaotiantian/viztracer#85]
Hi,

I'm trying to profile a multiprocessing project I'm working on, but the HTML output of VizTracer only shows the main process.
This is the case even with a simple multiprocessing program:
```python
import multiprocessing as mp
from time import sleep


def worker(rank):
    sleep(3)
    print(rank)


def main():
    worker(0)
    ctx = mp.get_context('spawn')
    processes = []
    for i in range(2):
        p = ctx.Process(target=worker, args=(i,))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()


if __name__ == '__main__':
    main()
```

To profile this, I'm using: `python -m viztracer --log_multiprocess profile_this.py`. Below is a screenshot of the HTML output:
![image](https://user-images.githubusercontent.com/77580049/105201016-0f28bb80-5b49-11eb-9736-35eb802d6b1e.png)

Here is also a zip file containing the `result.html` output: [result.zip](https://github.com/gaogaotiantian/viztracer/files/5843654/result.zip)
What's wrong here? Am I missing something?

**Comment by gaogaotiantian** [comment]
Maybe the documentation is not clear enough:

> This feature is only available on UNIX, not Windows. Also it will only work when ```multiprocessing.get_start_method()``` is ```fork```. After python3.8, MacOS has start method “spawn” by default and “fork” should be considered unsafe.

This does not mean you can use ```spawn``` context when the OS supports ```fork```. This means VizTracer only supports ```fork``` process. 

Normally VizTracer will complain if ```get_start_method()``` is not ```fork```. However you used ```get_context```, which I'm not sure how to detect, so VizTracer just let your program run without warning anything.

**Comment by gaogaotiantian** [comment]
Actually, I'm working on a prototype on spawned process and I'm making good progress. I think VizTracer will support spawned processes soon.

**Comment by rafi-cohen** [comment]
> Maybe the documentation is not clear enough:
> 
> > This feature is only available on UNIX, not Windows. Also it will only work when `multiprocessing.get_start_method()` is `fork`. After python3.8, MacOS has start method “spawn” by default and “fork” should be considered unsafe.

Huh, I guess I misunderstood that part. Thanks for clarifying.

> Actually, I'm working on a prototype on spawned process and I'm making good progress. I think VizTracer will support spawned processes soon.

That's great to hear! The project I'm working on heavily relies on a library which does not support `fork`, so this could make things easier for me.
For now I guess I'll try to run VizTracer inline within each process and then combine the results. Sounds like that should be possible.

**Comment by gaogaotiantian** [comment]
Yes, run VizTracer inline with ```pid_suffix=True``` and output file as ```json```, then combining them is always a way to do this. I'm working on spawn process and I think I could make it work.

**Comment by rafi-cohen** [comment]
Yes, that's exactly what I did.
I was able to identify the issue in my project immediately, thanks!

By the way, while using VizTracer, I noticed it didn't trace some parts of my project. I use [PyTorch's Distributed RPC Framework](https://pytorch.org/docs/stable/rpc.html) to set up a master process that handles requests from other processes. Basically, VizTracer does not trace the requests that are handled by the master process.
I'm not sure if this is a limitation that can be fixed, but thought I'd let you know about it.

If you wish to test this out, here's a simple code to reproduce:
```python
import os
from time import sleep
from torch.distributed import rpc
from torch import multiprocessing as mp


def do_something(t):
    print("Doing something")
    sleep(t)


def master(rank, world_size):
    rpc.init_rpc(name='master', rank=rank, world_size=world_size)
    rpc.shutdown()  # Handle requests until all processes shutdown


def worker(rank, world_size):
    rpc.init_rpc(name=f'worker{rank}', rank=rank, world_size=world_size)
    rpc.rpc_sync(to='master', func=do_something, args=(3,))  # Blocks until the master process finishes do_something(3)
    rpc.shutdown()


def deploy(rank, world_size):
    if rank == 0:
        master(rank, world_size)
    else:
        worker(rank, world_size)


def main():
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29500'
    world_size = 2  # one master, rest are workers
    mp.start_processes(deploy, args=(world_size,), nprocs=world_size, join=True, start_method='fork')


if __name__ == '__main__':
    main()
```
I used `fork` in this example, so tracing is done simply by using `python -m viztracer --log_multiprocess reproduce.py`
The worker makes a request to the master to run `do_something(3)`, so you'd expect to see `do_something` under the master process in the trace viewer, but it's nowhere to be found.

**Comment by gaogaotiantian** [comment]
```torch``` is too deep a hole to dig, but I looked at the code briefly. I guess it's because they are handling the rpc on C level. The master may have some rpc worker threads that are spawned directly using C(not ``threading`` module in python), which VizTracer can't catch and hook. 

However, I'm happy that VizTracer helped you for your project! Share with your friends when you have some time and if you want to :)

I'm still working on the spawn process tracing, I think I should be able to release it soon.

**Comment by gaogaotiantian** [comment]
Hi @rafi-cohen,

I want to thank you for using VizTracer and giving feedbacks to me. You gave me the motive to add ``spawn`` process support for VizTracer. After some hacking this feature is released with 0.11.3. If you want, please try it out and see if it works for you.

**Comment by rafi-cohen** [comment]
Ha! I should be the one thanking you, that's a great tool you made here.
Glad to hear that `spawn` is supported now. I just tested it, looks like it's working perfectly.

Kudos for working on this so fast!
